### PR TITLE
[CL][Incentives] Add authorized uptimes as a module param

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
Closes: #4915

## What is the purpose of the change

This PR adds `AuthorizedUptimes` as a module param to CL. Note that all core logic will continue to accommodate the full list of `SupportedUptimes` to avoid requiring a state migration in case governance wants to enable charging/authorize more uptimes. As such, `AuthorizedUptimes` will always be a subset of `SupportedUptimes`, with a default value of just supporting 1 nanosecond (equivalent to status quo incentives.

Since accumulator updates are skipped when accums are empty, the performance overhead for continuing to accommodate these uptimes should be minimal.

## Brief Changelog

- Add `AuthorizedUptimes` as a module param
- Modify incentive creation logic to check against `AuthorizedUptimes` instead of `SupportedUptimes`
- Update incentive creation tests to test a range of authorized uptimes

## Testing and Verifying

- Changes are tested in `incentives_test.go`

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not documented)

### CODE REVIEW

File: x/concentrated-liquidity/incentives_test.go
Comment (AlpinYukseloglu) [inline x/concentrated-liquidity/incentives_test.go:2915]: Note to reviewer: we add this field to test a range of supported uptimes and, importantly, to ensure that existing test cases do not pass or fail trivially due to only authorizing 1 nanosecond uptime after disabling charging

File: x/incentives/keeper/distribute_test.go
Comment (czarcas7ic) [inline x/incentives/keeper/distribute_test.go:198]: Every test case here defines authorizedUptimes as the supportedUptimes. Do you want to add some other test cases that do not touch this branch? If not / you do not see any value in it, we can just remove the param from the test case entirely and set it on every test run here.

File: x/incentives/keeper/distribute_test.go
Comment (czarcas7ic) [inline x/incentives/keeper/distribute_test.go:755]: Same comment here

File: x/concentrated-liquidity/incentives.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives.go:864]: I think we can break here on setting to true to save on iterations

File: x/concentrated-liquidity/types/params.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/types/params.go:187]: I think we can also break here to save on unnecessary iterations.

**Review by czarcas7ic — APPROVED** [review]
Sorry my review got split up for some reason when I was reviewing via vscode. All looks good to me with minor nits!

File: x/incentives/keeper/distribute_test.go
Comment (AlpinYukseloglu) [reply x/incentives/keeper/distribute_test.go:215]: I had originally tabled this to avoid touching gauge logic per this issue #4985, but the change was small enough where I just tied it into this PR. Wider range of authorized uptimes (including default of 1ns) should be tested now.

File: x/incentives/keeper/distribute_test.go
Comment (AlpinYukseloglu) [reply x/incentives/keeper/distribute_test.go:785]: See c31100c75598a516e5b3106140f3b4c0654549f5 and ec471d648016de079af19f01b52771e6e5f98c39

File: x/concentrated-liquidity/incentives.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/incentives.go:864]: Good point!

Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/incentives.go:864]: 4060398207ce4038a9ab1307c30180f62b75aa17

File: x/concentrated-liquidity/types/params.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/types/params.go]: 4060398207ce4038a9ab1307c30180f62b75aa17

**Comment by p0mvn** [comment]
Please make sure that when you run:
`make localnet-start-with-state`, the script does not break:
https://github.com/osmosis-labs/osmosis/tree/main/tests/cl-genesis-positions

The way to check is to confirm that there is a CL pool with id 1 and many positions created upon running `make localnet-start-with-state`

From what I can tell, there is one update that is necessary atm:
```bash
# from project root
cd tests/cl-genesis-positions
go get github.com/osmosis-labs/osmosis/v15@4060398
# observe go mod updating, commit it
```

Let me know if you need help with this but please do not merge, unless confirmed that the script still works to avoid breaking integrator testing

File: x/concentrated-liquidity/incentives.go
Comment (p0mvn) [inline x/concentrated-liquidity/incentives.go:859]: What is the purpose of sorting?

**Review by p0mvn — APPROVED** [review]
LGTM upon testing localosmosis script not breaking

Also, please add a test for checking the param in the upgrade handler by adding a validation below these lines:
https://github.com/osmosis-labs/osmosis/blob/85a748062dce94efaacb488ffcca6c9d25c5e588/app/upgrades/v16/upgrades_test.go#L99-L101

File: x/concentrated-liquidity/incentives.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/incentives.go:859]: My assumption is that uptimes can technically be stored in non-increasing order, which I believe we might have some assumptions leaning on (even if just in tests). This was intended to be a low-cost guardrail to ensure that ordering is not an issue. Do you see an issue with this reasoning?

**Comment by p0mvn** [comment]
@AlpinYukseloglu FYI starting to work on localosmosis, will merge once completed

**Comment by p0mvn** [comment]
Tested localosmosis, it passes now. Should be good to merge as long as CI passes

**Pull request merged**
